src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -884,7 +882,8 @@ ZooKeeperPtr ZooKeeper::startNewSession() const
 
 bool ZooKeeper::expired()
 {
-    return impl->isExpired();
+    bool fallback_session_expired = ((session_lifetime_seconds > 0) && (getSessionUptime() >= session_lifetime_seconds));
+    return fallback_session_expired || impl->isExpired();
```

Comment:
The only correct way to expire the current session and start a new one is to call `finalize`. Otherwise, some objects will start using a new session while others still use the old one. It will cause consistency issues and will trigger obscure bugs. See https://github.com/ClickHouse/ClickHouse/pull/50424#issuecomment-1611812850